x/cron/keeper/msg_server.go

```diff
@@ -0,0 +1,107 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dymensionxyz/rollapp-wasm/x/cron/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (k msgServer) RegisterContract(goCtx context.Context, msg *types.MsgRegisterContract) (*types.MsgRegisterContractResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	// check if the cron is globally enabled
+	params := k.GetParams(ctx)
+	if !params.EnableCron {
+		return &types.MsgRegisterContractResponse{}, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Cron is disabled")
+	}
+
+	if err := msg.ValidateBasic(); err != nil {
+		ctx.Logger().Error(fmt.Sprintf("request invalid: %s", err))
+		return &types.MsgRegisterContractResponse{}, err
+	}
+
+	// Validation such that only the user who instantiated the contract can register contract
+	contractAddr, err := sdk.AccAddressFromBech32(msg.ContractAddress)
+	if err != nil {
+		return &types.MsgRegisterContractResponse{}, sdkerrors.ErrInvalidAddress
+	}
+	contractInfo := k.conOps.GetContractInfo(ctx, contractAddr)
+
+	// check if sender is authorized
+	exists := k.CheckSecurityAddress(ctx, msg.SecurityAddress)
+	if !exists {
+		return &types.MsgRegisterContractResponse{}, sdkerrors.ErrUnauthorized
+	}
+
+	allContracts := k.GetWhitelistedContracts(ctx)
+
+	for _, data := range allContracts {
+		if data.ContractAddress == msg.ContractAddress {
+			return &types.MsgRegisterContractResponse{}, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "contract already registered")
+		}
+	}
+	gameID := k.GetGameID(ctx)
+	contract := types.WhitelistedContract{
+		GameId:          gameID + 1,
```

Comment:
+1 needs explanation

Reply:
gameID is a counter which is incremented whenever a new game contract is whitelisted.